backend/main/chapters/c09_combining_booleans.py

```diff
@@ -0,0 +1,638 @@
+let's explore how `and` and `or` interact when combined, using 3 booleans.
+Which one has higher priority? Try the following code in the shell.
+What do you expect?
+
+__program_indented__
+        """
+        program = "True or False and False"
+
+    class AndHasHigherPriority(VerbatimStep):
+        """
+If you read it casually from left to right, you may think:
+
+- `True or False` would evaluate to `True`,
+- then it would reduce to `True and False`, which evaluates to `False`.
+
+However, this is not the case! Because `and` has a higher priority in evaluation than `or`, so
+
+    True or False and False
+
+is equivalent to
+
+    True or (False and False)
+
+which evaluates to `True` (what's in parentheses will be evaluated first).
+
+So we should always use parentheses to be safe.
+
+Now evaluate in the shell the command:
+
+__program_indented__
+        """
+        program = "(True or False) and False"
+
+    final_text = """
+Therefore we should play it safe and always use parentheses to indicate
+which priority we intend with our code when there is ambiguity.
+    """
+
+
+class MultiLineExpressions(Page):
+    class TrueOrFalseAndFalse(VerbatimStep):
+        """
+Our code for checking a tic-tac-toe board for a winning row will get quite long.
+Thankfully Python allows us to write a single, long expression on multiple lines,
+as long as it is within parentheses.
+We can go to the next line right after a boolean operator.
+Type and run this code in the editor:
+
+    print(True or
+    False and
+    False)
+        """
+        program_in_text = False
+        program = "print(True or False and False)"
+
+    final_text = """
+Remember to always wrap your multi-line expressions inside parentheses to be safe.
+    """
+
+
+class CombiningAndWithOr(Page):
+    class CheckTheWholeBoard(VerbatimStep):
+        """
+The tic-tac-toe board can be represented by 3 rows.
+If we have a winning row, it's either row 1, OR row 2, OR row 3.
+For each row, we can reuse our row-checking code from before
+(with parentheses around it):
+
+    (row[0] == row[1] and row[0] == row[2])
+
+To fit all 3 row checks on multiple lines,
+we need one extra set of parentheses around the whole thing, after `if`.
+We break lines after the boolean operators `or` which are connecting the bigger blocks.
+Type and run this code, make sure you understand the complex boolean expression:
+
+__program_indented__
+        """
+        def program(self):
+            def is_winning_board(row1, row2, row3):
+                if (
+                    (row1[0] == row1[1] and row1[0] == row1[2]) or
+                    (row2[0] == row2[1] and row2[0] == row2[2]) or
+                    (row3[0] == row3[1] and row3[0] == row3[2])
+                ):
+                    print("We have a winner!")
+
+            row1 = ["X", "O", "X"]
+            row2 = ["X", "X", "X"]
+            row3 = ["O", "O", "X"]
+
+            is_winning_board(row1, row2, row3)
```

Comment:
Now that we have functions, this has become a bad example. We're showing them bad code. This should be written as:

```python
def is_winning_row(row):
    return row[0] == row[1] and ...

def is_winning_board(row1, row2, row3):
    return is_winning_row(row1) or ...
```

I think we need to think of a new example.

Reply:
One easy cheat is to change the exercise to find winning Columns instead.

Reply:
No. Eventually we will have introduce nested lists before this chapter and the examples will be modified to reflect that instead of having three lists. Then it'll be `def is_winning_column(board, column): return board[0][column] == ...`.

I think it's actually fine to skip to the exercise with diagonals. We've shown them the main pitfall and they have plenty of tools to help them. I no longer think we need to be so thorough, this isn't like loops.

Reply:
OK! That makes my life easier! Will make the changes and commit.